File: javascript/collatz-conjecture/collatz-conjecture.js
//
// This is only a SKELETON file for the 'Collatz Conjecture' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const steps = (number) => {
  if (number <= 0) {
    throw new Error("Only positive numbers are allowed");
  }
  const iter = (acc, num) => {
    if (num === 1) {
      return acc;
    }
    return iter(acc + 1, num % 2 === 0 ? num / 2 : num * 3 + 1);
  };
  
  return iter(0, number);
};

File: python/collatz-conjecture/collatz_conjecture.py
def steps(number):
    if number < 1:
        raise ValueError('Number must be positive')

    def iter(num, count):
        if num == 1:
            return count
        if num % 2 == 0:
            return iter(num / 2, count + 1)
        return iter(num * 3 + 1, count + 1)

    return iter(number, 0)
